File: backend/app/schemas/product.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict
from enum import Enum

class ProductCondition(str, Enum):
    NEW = "new"
    USED_LIKE_NEW = "used_like_new"
    USED_GOOD = "used_good"
    USED_FAIR = "used_fair"

class ProductStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    INACTIVE = "inactive"
    OUT_OF_STOCK = "out_of_stock"

class SizeCreate(BaseModel):
    size: str
    stock: int = 0
    price_adjustment: float = 0.0

class ColorCreate(BaseModel):
    name: str
    color_code: str
    images: List[str] = []
    sizes: List[SizeCreate] = []

class ProductVariantCreate(BaseModel):
    sku: str
    color: Optional[str] = None
    size: Optional[str] = None
    stock: int = 0
    price: float
    sale_price: Optional[float] = None
    images: List[str] = []

class ShippingDetailsCreate(BaseModel):
    weight: Optional[float] = None
    dimensions: Optional[Dict[str, float]] = None
    shipping_cost: Optional[float] = None
    free_shipping: bool = False

class SEOCreate(BaseModel):
    meta_title: Optional[str] = None
    meta_description: Optional[str] = None
    keywords: List[str] = []

class ProductCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=200)
    name_ar: Optional[str] = None
    description: str = Field(..., min_length=10)
    description_ar: Optional[str] = None
    
    # Required fields
    boutique_id: str
    category: str
    base_price: float = Field(..., gt=0)
    main_image: str
    
    # Optional fields
    subcategory: Optional[str] = None
    tags: List[str] = []
    brand: Optional[str] = None
    model: Optional[str] = None
    sale_price: Optional[float] = None
    currency: str = "DZD"
    
    # Variants & Stock
    variants: List[ProductVariantCreate] = []
    colors: List[ColorCreate] = []
    total_stock: int = 0
    min_stock_alert: int = 5
    
    # Product Details
    condition: ProductCondition = ProductCondition.NEW
    material: Optional[str] = None
    care_instructions: Optional[str] = None
    
    # Images & Media
    images: List[str] = []
    video_url: Optional[str] = None
    
    # Shipping
    shipping_details: ShippingDetailsCreate = ShippingDetailsCreate()
    
    # Status
    status: ProductStatus = ProductStatus.ACTIVE
    is_featured: bool = False
    is_trending: bool = False
    
    # SEO
    seo: SEOCreate = SEOCreate()
    
    @validator("sale_price")
    def validate_sale_price(cls, v, values):
        if v is not None and "base_price" in values and v >= values["base_price"]:
            raise ValueError("Sale price must be less than base price")
        return v

class ProductUpdate(BaseModel):
    name: Optional[str] = None
    name_ar: Optional[str] = None
    description: Optional[str] = None
    description_ar: Optional[str] = None
    category: Optional[str] = None
    subcategory: Optional[str] = None
    tags: Optional[List[str]] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    base_price: Optional[float] = None
    sale_price: Optional[float] = None
    total_stock: Optional[int] = None
    condition: Optional[ProductCondition] = None
    material: Optional[str] = None
    care_instructions: Optional[str] = None
    status: Optional[ProductStatus] = None
    is_featured: Optional[bool] = None
    is_trending: Optional[bool] = None

class ProductFilters(BaseModel):
    category: Optional[str] = None
    subcategory: Optional[str] = None
    boutique_id: Optional[str] = None
    brand: Optional[str] = None
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    color: Optional[str] = None
    size: Optional[str] = None
    condition: Optional[ProductCondition] = None
    status: Optional[ProductStatus] = ProductStatus.ACTIVE
    is_featured: Optional[bool] = None
    is_trending: Optional[bool] = None
    in_stock_only: bool = True
    search: Optional[str] = None

class ProductSort(str, Enum):
    RELEVANCE = "relevance"
    PRICE_ASC = "price_asc"
    PRICE_DESC = "price_desc"
    NEWEST = "newest"
    OLDEST = "oldest"
    POPULARITY = "popularity"
    RATING = "rating"
    SALES = "sales"

class ProductListParams(BaseModel):
    page: int = Field(1, ge=1)
    size: int = Field(20, ge=1, le=100)
    sort: ProductSort = ProductSort.RELEVANCE
    filters: ProductFilters = ProductFilters()

class ProductResponse(BaseModel):
    id: str
    name: str
    name_ar: Optional[str]
    slug: str
    description: str
    description_ar: Optional[str]
    boutique_id: str
    boutique_name: str
    category: str
    subcategory: Optional[str]
    tags: List[str]
    brand: Optional[str]
    base_price: float
    sale_price: Optional[float]
    current_price: float
    discount_percentage: Optional[float]
    currency: str
    main_image: str
    images: List[str]
    condition: ProductCondition
    status: ProductStatus
    is_featured: bool
    is_trending: bool
    is_in_stock: bool
    total_stock: int
    rating: float
    rating_count: int
    sales_count: int
    views: int
    created_at: str
    
    class Config:
        from_attributes = True

class ProductListResponse(BaseModel):
    products: List[ProductResponse]
    total: int
    page: int
    size: int
    total_pages: int
    has_next: bool
    has_prev: bool

class ProductDetailResponse(ProductResponse):
    variants: List[dict]
    colors: List[dict]
    material: Optional[str]
    care_instructions: Optional[str]
    shipping_details: dict
    seo: dict
    available_sizes: List[str]
    available_colors: List[str]
    
class CategoryResponse(BaseModel):
    name: str
    slug: str
    product_count: int
    subcategories: List[str] = []

class BrandResponse(BaseModel):
    name: str
    product_count: int
    logo: Optional[str] = None

File: frontend/src/services/product.ts
import api from './auth';

export interface ProductFilters {
  category?: string;
  subcategory?: string;
  boutique_id?: string;
  brand?: string;
  min_price?: number;
  max_price?: number;
  color?: string;
  size?: string;
  condition?: string;
  search?: string;
  in_stock_only?: boolean;
  is_featured?: boolean;
  is_trending?: boolean;
}

export interface Product {
  id: string;
  name: string;
  name_ar?: string;
  slug: string;
  description: string;
  description_ar?: string;
  boutique_id: string;
  boutique_name: string;
  category: string;
  subcategory?: string;
  tags: string[];
  brand?: string;
  base_price: number;
  sale_price?: number;
  current_price: number;
  discount_percentage?: number;
  currency: string;
  main_image: string;
  images: string[];
  condition: string;
  status: string;
  is_featured: boolean;
  is_trending: boolean;
  is_in_stock: boolean;
  total_stock: number;
  rating: number;
  rating_count: number;
  sales_count: number;
  views: number;
  created_at: string;
}

export interface ProductDetail extends Product {
  variants: any[];
  colors: any[];
  material?: string;
  care_instructions?: string;
  shipping_details: any;
  seo: any;
  available_sizes: string[];
  available_colors: string[];
}

export interface ProductListResponse {
  products: Product[];
  total: number;
  page: number;
  size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface Category {
  name: string;
  slug: string;
  product_count: number;
  subcategories: string[];
}

export interface Brand {
  name: string;
  product_count: number;
  logo?: string;
}

export const productService = {
  async getProducts(
    page = 1,
    size = 20,
    sort = 'relevance',
    filters: ProductFilters = {}
  ): Promise<ProductListResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sort,
      ...Object.fromEntries(
        Object.entries(filters).filter(([_, value]) => value !== undefined && value !== '')
      )
    });

    const response = await api.get(`/products?${params}`);
    return response.data;
  },

  async getProduct(id: string): Promise<ProductDetail> {
    const response = await api.get(`/products/${id}`);
    return response.data;
  },

  async getProductBySlug(slug: string): Promise<ProductDetail> {
    const response = await api.get(`/products/slug/${slug}`);
    return response.data;
  },

  async getSimilarProducts(productId: string, limit = 8): Promise<Product[]> {
    const response = await api.get(`/products/${productId}/similar?limit=${limit}`);
    return response.data;
  },

  async getFeaturedProducts(limit = 10): Promise<Product[]> {
    const response = await api.get(`/products/featured?limit=${limit}`);
    return response.data;
  },

  async getTrendingProducts(limit = 10): Promise<Product[]> {
    const response = await api.get(`/products/trending?limit=${limit}`);
    return response.data;
  },

  async getNewArrivals(limit = 10): Promise<Product[]> {
    const response = await api.get(`/products/new-arrivals?limit=${limit}`);
    return response.data;
  },

  async getCategories(): Promise<Category[]> {
    const response = await api.get('/products/categories');
    return response.data;
  },

  async getBrands(): Promise<Brand[]> {
    const response = await api.get('/products/brands');
    return response.data;
  },

  async createProduct(data: any): Promise<Product> {
    const response = await api.post('/products', data);
    return response.data;
  },

  async updateProduct(id: string, data: any): Promise<Product> {
    const response = await api.put(`/products/${id}`, data);
    return response.data;
  },

  async deleteProduct(id: string): Promise<void> {
    await api.delete(`/products/${id}`);
  },
};
